google_guest_agent/snapshot_listener.go

```diff
@@ -0,0 +1,252 @@
+				logger.Errorf("error reading request: %v", err)
+				break
+			}
+			logger.Infof("received snapshot request")
+			requestChan <- request
+		}
+		cancel()
+	}
+}
+
+func getSnapshotResponse(guestMessage *pb.GuestMessage) *pb.SnapshotResponse {
+	switch request := guestMessage.Msg.(type) {
+	case *pb.GuestMessage_SnapshotRequest:
+		response := &pb.SnapshotResponse{
+			OperationId: request.SnapshotRequest.GetOperationId(),
+			Type:        request.SnapshotRequest.GetType(),
+		}
+
+		config, err := getSnapshotConfig()
+		if err != nil {
+			response.AgentReturnCode = pb.AgentErrorCode_INVALID_CONFIG
+			return response
+
+		}
+
+		var url string
+		switch request.SnapshotRequest.GetType() {
+		case pb.OperationType_PRE_SNAPSHOT:
+			logger.Infof("handling pre snapshot request for operation id %d", request.SnapshotRequest.GetOperationId())
+			if seenPreSnapshotoperationIds.Contains(request.SnapshotRequest.GetOperationId()) {
+				logger.Infof("duplicate pre snapshot request operation id %d", request.SnapshotRequest.GetOperationId())
+				return nil
+			}
+			seenPreSnapshotoperationIds.Add(request.SnapshotRequest.GetOperationId(), nil)
+			url = config.PreSnapshotScriptURL
+		case pb.OperationType_POST_SNAPSHOT:
+			logger.Infof("handling post snapshot request for operation id %d", request.SnapshotRequest.GetOperationId())
+			if seenPostSnapshotoperationIds.Contains(request.SnapshotRequest.GetOperationId()) {
+				logger.Infof("duplicate post snapshot request operation id %d", request.SnapshotRequest.GetOperationId())
+				return nil
+			}
+			seenPostSnapshotoperationIds.Add(request.SnapshotRequest.GetOperationId(), nil)
+			url = config.PostSnapshotScriptURL
+		default:
+			logger.Errorf("unhandled operation type %d", request.SnapshotRequest.GetType())
+			return nil
+		}
+
+		agentErrorCode, scriptsReturnCode := handleRequest(config, url, request.SnapshotRequest.GetDiskList())
+		response.ScriptsReturnCode = int32(scriptsReturnCode)
+		response.AgentReturnCode = agentErrorCode
+
+		return response
+	case *pb.GuestMessage_ServerInfo:
+	}
+	return nil
+}
+
+func handleSnapshotRequests(address string, requestChan chan *pb.GuestMessage) {
+	for {
+		conn, err := grpc.Dial(address, grpc.WithInsecure(), grpc.WithBlock())
+		if err != nil {
+			logger.Errorf("failed to connect: %v", err)
+			continue
+		}
+		for {
+			// Listen on channel and respond
+			guestMessage := <-requestChan
+			response := getSnapshotResponse(guestMessage)
+			for {
+				c := pb.NewSnapshotServiceClient(conn)
+				ctx, cancel := context.WithCancel(context.Background())
+				_, err = c.HandleResponsesFromGuest(ctx, response)
+				if err == nil {
+					cancel()
+					break
+				}
+				logger.Errorf("error sending response: %v", err)
+			}
+		}
+	}
+
+}
+
+func startSnapshotListener(testMode bool, testSnapshotPort int) {
+	requestChan := make(chan *pb.GuestMessage)
+	var address string
+	if testMode {
+		address = testIP + ":" + strconv.Itoa(testSnapshotPort)
```

Comment:
`fmt.Sprintf` seems friendlier

Reply:
this is also probably obviated by the feedback on test constants

Reply:
Holding off on this for now